crates/router/src/core/payments/tokenization.rs

```diff
@@ -780,62 +780,79 @@ pub async fn add_payment_method_token<F: Clone, T: types::Tokenizable + Clone>(
+                    .set_session_token(pm_token_router_data.session_token.clone());
+
+                let resp = services::execute_connector_processing_step(
+                    state,
+                    connector_integration,
+                    &pm_token_router_data,
+                    payments::CallConnectorAction::Trigger,
+                    None,
+                )
+                .await
+                .to_payment_failed_response()?;
+
+                metrics::CONNECTOR_PAYMENT_METHOD_TOKENIZATION.add(
+                    &metrics::CONTEXT,
+                    1,
+                    &add_attributes([
+                        ("connector", connector.connector_name.to_string()),
+                        ("payment_method", router_data.payment_method.to_string()),
+                    ]),
                 );
 
-            router_data
-                .request
-                .set_session_token(pm_token_router_data.session_token.clone());
-
-            let resp = services::execute_connector_processing_step(
-                state,
-                connector_integration,
-                &pm_token_router_data,
-                payments::CallConnectorAction::Trigger,
-                None,
-            )
-            .await
-            .to_payment_failed_response()?;
-
-            metrics::CONNECTOR_PAYMENT_METHOD_TOKENIZATION.add(
-                &metrics::CONTEXT,
-                1,
-                &add_attributes([
-                    ("connector", connector.connector_name.to_string()),
-                    ("payment_method", router_data.payment_method.to_string()),
-                ]),
-            );
-
-            let pm_token = match resp.response {
-                Ok(response) => match response {
-                    types::PaymentsResponseData::TokenizationResponse { token } => Some(token),
-                    _ => None,
-                },
-                Err(err) => {
-                    logger::debug!(payment_method_tokenization_error=?err);
-                    None
-                }
-            };
-            Ok(pm_token)
+                let (pm_token, should_continue_further) = match resp.response {
+                    Ok(response) => match response {
+                        types::PaymentsResponseData::TokenizationResponse { token } => {
+                            (Some(token), should_continue_payment)
+                        }
+                        _ => (None, should_continue_payment),
+                    },
+                    Err(err) => {
+                        if is_retry_payment {
+                            router_data.response = Err(err);
+                            println!("failed connector tokenization");
+                            (None, false)
+                        } else {
+                            logger::debug!(payment_method_tokenization_error=?err);
+                            (None, should_continue_payment)
+                        }
```

Comment:
you can decouple this logic as it is done for access token, this function only returns the result of payment method tokenization. Rest of the decision can be taken by another function. You can try to do this.